### ISSUE TITLE
Support nested groups

### ISSUE THREAD

**Issue by iainhallam** [splitbrain/dokuwiki#1833]
I'm trying to model a permission structure in our organisation that has users assigned to groups as role holders, and then those groups are assigned and aggregated to other groups for permission to do things. This makes things much easier to manage than to just have to either assign individual users to every group or to have a huge proliferation of groups to represent the hierarchical nature of the permissions.

Looking at the code, I guess the auth_aclcheck_cb function would need changing, and some UI in the admin manager, either in the existing user manager plugin or as a group manager plugin. I suspect the auth backends could be left alone, as long as the core knows to expand any groups before using them.

**Comment by splitbrain** [comment]
Hmm. First of all you should not modify ``auth_aclcheck_cb`` but write a plugin handling the ``AUTH_ACL_CHECK`` event.

Nested groups just means that I'm automatically a member of all parent groups of my actual groups, right? I guess this could also be hacked into DokuWiki using one of the earlier events. Maybe ``AUTH_LOGIN_CHECK``. Whenever a  user logs in, add the parent groups to their group array....

**Comment by iainhallam** [comment]
I'm aware I can write a plugin, but I was opening a bug to report this against DokuWiki itself - I'd like the core to support memberships of groups constructed like that, so that any auth plugin can make use of the feature, instead of having to reimplement it.

Yes, it effectively means you become a member of all the parent groups. Taking that interpretation literally, though, might lead to confusion in auth plugins when given a list of groups to write for a user modification. I thought from the look of it that the authplain plugin would suddenly be writing all the dereferenced groups beck into the user line, for instance.

**Comment by splitbrain** [comment]
Yes, it effectively means you become a member of all the parent groups.
Taking that interpretation literally, though, might lead to confusion in
auth plugins when given a list of groups to write for a user modification.
I thought from the look of it that the authplain plugin would suddenly be
writing all the dereferenced groups beck into the user line, for instance.

Ah, true. I guess using the ACL check makes more sense then. There might
still be problems with how the ACL manager predicts permissions then...

Re: core support. I would want to have nested groups in core. ACLs are
complicated and confusing enough as is, but good enough to cover most
user's needs. An action plugin seems to me the most reasonable approach.
